File: src/Core.TypeScript/chip8/channel-grant.ts
import { assistedRunChannelLabel, keyText, type RunChannelLabel, type RunKey } from "../chip9/chip8-cross-run-store";

/**
 * Honest boundary: TypeScript cannot make the harness factory inaccessible to another module,
 * and in-process code can still touch `frame.mem` directly. The branded token plus WeakMap gives
 * the typed path a runtime refusal; process or WASM isolation is required for a stronger claim.
 */

export type ChannelDirection = "read" | "write";

export interface ChannelSpec {
  readonly channel: string;
  readonly direction: ChannelDirection;
  readonly startAddress: number;
  readonly endAddress: number;
}

export type ChannelGrantFeedbackCode =
  | "empty-channel-set"
  | "invalid-channel-name"
  | "invalid-channel-range"
  | "overlapping-channel-ranges"
  | "invalid-experimenter-id"
  | "invalid-channel-set"
  | "invalid-derived-channel-label"
  | "run-key-channel-mismatch"
  | "invalid-channel-grant"
  | "crossing-not-granted"
  | "crossing-count-overflow"
  | "run-key-digest-failed";

export interface ChannelGrantFeedback {
  readonly code: ChannelGrantFeedbackCode;
  readonly detail: string;
}

export type ChannelGrantResult<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly feedback: ChannelGrantFeedback };

const fail = <T>(code: ChannelGrantFeedbackCode, detail: string): ChannelGrantResult<T> => ({
  ok: false,
  feedback: { code, detail },
});

const channelSetBrand: unique symbol = Symbol("zeta.channel-set");
const grantBrand: unique symbol = Symbol("zeta.channel-grant");

export interface ChannelSet {
  readonly [channelSetBrand]: true;
  readonly channels: readonly ChannelSpec[];
}

export interface ChannelGrant {
  readonly [grantBrand]: true;
  readonly channels: ChannelSet;
  readonly issuedBy: string;
  readonly runKey: Readonly<RunKey>;
  readonly channelLabel: RunChannelLabel;
}

export interface ChannelCrossing {
  readonly channel: string;
  readonly direction: ChannelDirection;
  readonly address: number;
}

export interface ChannelMeterRow extends ChannelSpec {
  readonly crossings: number;
}

export interface ChannelMeterSnapshot {
  readonly channelLabel: RunChannelLabel;
  readonly issuedBy: string;
  readonly runKey: string;
  readonly rows: readonly ChannelMeterRow[];
}

export interface ChannelGrantHarness {
  readonly issuedBy: string;
  issue(runKey: RunKey, channels: ChannelSet): ChannelGrantResult<ChannelGrant>;
}

interface GrantState {
  readonly grant: ChannelGrant;
  readonly counts: Map<string, number>;
}

const issuedChannelSets = new WeakSet<object>();
const issuedGrants = new WeakMap<object, GrantState>();
const MAX_ADDRESS = 0xfff;

/** The complete apparatus used by the two live memory-observing CHIP-8 hosts. */
export const FULL_RAM_TAS_CHANNELS: readonly ChannelSpec[] = Object.freeze([
  Object.freeze({ channel: "ram", direction: "read", startAddress: 0, endAddress: MAX_ADDRESS }),
  Object.freeze({ channel: "ram", direction: "write", startAddress: 0, endAddress: MAX_ADDRESS }),
]);

function directionText(direction: ChannelDirection): string {
  return direction;
}

function canonicalSpec(spec: ChannelSpec): string {
  return `${spec.channel}-${directionText(spec.direction)}@${spec.startAddress.toString(16).padStart(4, "0")}-${spec.endAddress.toString(16).padStart(4, "0")}`;
}

function ordinal(left: string, right: string): number {
  if (left < right) return -1;
  if (left > right) return 1;
  return 0;
}

function rangesOverlap(left: ChannelSpec, right: ChannelSpec): boolean {
  return (
    left.channel === right.channel &&
    left.direction === right.direction &&
    left.startAddress <= right.endAddress &&
    right.startAddress <= left.endAddress
  );
}

function isChannelSet(value: ChannelSet): boolean {
  return issuedChannelSets.has(value);
}

export function createChannelSet(specs: readonly ChannelSpec[]): ChannelGrantResult<ChannelSet> {
  if (specs.length === 0) return fail("empty-channel-set", "at least one TAS channel is required");

  const channels = specs.map((spec) => ({ ...spec }));
  for (const spec of channels) {
    if (!/^[a-z][a-z0-9-]*$/.test(spec.channel)) {
      return fail("invalid-channel-name", spec.channel);
    }
    if (
      !Number.isInteger(spec.startAddress) ||
      !Number.isInteger(spec.endAddress) ||
      spec.startAddress < 0 ||
      spec.endAddress > MAX_ADDRESS ||
      spec.startAddress > spec.endAddress
    ) {
      return fail("invalid-channel-range", `${spec.channel}:${String(spec.startAddress)}-${String(spec.endAddress)}`);
    }
  }

  for (let left = 0; left < channels.length; left += 1) {
    for (let right = left + 1; right < channels.length; right += 1) {
      const a = channels[left];
      const b = channels[right];
      if (a !== undefined && b !== undefined && rangesOverlap(a, b)) {
        return fail("overlapping-channel-ranges", `${a.channel}:${a.direction}`);
      }
    }
  }

  channels.sort((left, right) => ordinal(canonicalSpec(left), canonicalSpec(right)));
  channels.forEach(Object.freeze);
  const channelSet: ChannelSet = Object.freeze({
    [channelSetBrand]: true as const,
    channels: Object.freeze(channels),
  });
  issuedChannelSets.add(channelSet);
  return { ok: true, value: channelSet };
}

export function channelLabelFor(channels: ChannelSet): ChannelGrantResult<RunChannelLabel> {
  if (!isChannelSet(channels)) return fail("invalid-channel-set", "channel set was not created by createChannelSet");
  const label = assistedRunChannelLabel(channels.channels.map(canonicalSpec).join(","));
  return label.ok ? label : fail("invalid-derived-channel-label", `${label.feedback.code}:${label.feedback.detail}`);
}

function validExperimenterId(issuedBy: string): boolean {
  return /^[A-Za-z0-9._-]+$/.test(issuedBy);
}

export function createChannelGrantHarness(issuedBy: string): ChannelGrantResult<ChannelGrantHarness> {
  if (!validExperimenterId(issuedBy)) return fail("invalid-experimenter-id", issuedBy);

  return {
    ok: true,
    value: Object.freeze({
      issuedBy,
      issue(runKey: RunKey, channels: ChannelSet): ChannelGrantResult<ChannelGrant> {
        const label = channelLabelFor(channels);
        if (!label.ok) return label;
        if (runKey.channelLabel !== label.value) {
          return fail("run-key-channel-mismatch", `expected=${label.value};actual=${runKey.channelLabel}`);
        }

        const frozenRunKey = Object.freeze({ ...runKey });
        const grant: ChannelGrant = Object.freeze({
          [grantBrand]: true as const,
          channels,
          issuedBy,
          runKey: frozenRunKey,
          channelLabel: label.value,
        });
        issuedGrants.set(grant, {
          grant,
          counts: new Map(channels.channels.map((spec) => [canonicalSpec(spec), 0])),
        });
        return { ok: true, value: grant };
      },
    }),
  };
}

/** Source-owned harness edge: bind a grant to the exact ROM and apparatus before play begins. */
export async function issueChip8ChannelGrant(
  issuedBy: string,
  rom: Uint8Array,
  specs: readonly ChannelSpec[],
  seed: number,
): Promise<ChannelGrantResult<ChannelGrant>> {
  const channels = createChannelSet(specs);
  if (!channels.ok) return channels;
  const channelLabel = channelLabelFor(channels.value);
  if (!channelLabel.ok) return channelLabel;
  const harness = createChannelGrantHarness(issuedBy);
  if (!harness.ok) return harness;

  try {
    const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", rom.slice().buffer));
    const romSha256 = [...digest].map((byte) => byte.toString(16).padStart(2, "0")).join("");
    const runKey: RunKey = {
      romSha256,
      seedHex: BigInt(seed >>> 0)
        .toString(16)
        .padStart(16, "0"),
      loadAddrHex: "0200",
      dialect: "chip8",
      channelLabel: channelLabel.value,
      stepMapVersion: "chip8cow-step-v1",
    };
    return harness.value.issue(runKey, channels.value);
  } catch (error: unknown) {
    return fail("run-key-digest-failed", error instanceof Error ? error.message : String(error));
  }
}

function stateFor(grant: ChannelGrant): ChannelGrantResult<GrantState> {
  const state = issuedGrants.get(grant);
  return state === undefined
    ? fail("invalid-channel-grant", "grant was not issued by createChannelGrantHarness")
    : { ok: true, value: state };
}

function specFor(
  state: GrantState,
  channel: string,
  direction: ChannelDirection,
  address: number,
): ChannelSpec | undefined {
  return state.grant.channels.channels.find(
    (spec) =>
      spec.channel === channel &&
      spec.direction === direction &&
      address >= spec.startAddress &&
      address <= spec.endAddress,
  );
}

function snapshot(state: GrantState): ChannelMeterSnapshot {
  return Object.freeze({
    channelLabel: state.grant.channelLabel,
    issuedBy: state.grant.issuedBy,
    runKey: keyText(state.grant.runKey),
    rows: Object.freeze(
      state.grant.channels.channels.map((spec) =>
        Object.freeze({ ...spec, crossings: state.counts.get(canonicalSpec(spec)) ?? 0 }),
      ),
    ),
  });
}

export function channelMeterSnapshot(grant: ChannelGrant): ChannelGrantResult<ChannelMeterSnapshot> {
  const state = stateFor(grant);
  return state.ok ? { ok: true, value: snapshot(state.value) } : state;
}

export function meterCrossings(
  grant: ChannelGrant,
  crossings: readonly ChannelCrossing[],
): ChannelGrantResult<ChannelMeterSnapshot> {
  const state = stateFor(grant);
  if (!state.ok) return state;

  const increments = new Map<string, number>();
  for (const crossing of crossings) {
    if (!Number.isInteger(crossing.address)) {
      return fail("crossing-not-granted", `${crossing.channel}:${crossing.direction}@${String(crossing.address)}`);
    }
    const spec = specFor(state.value, crossing.channel, crossing.direction, crossing.address);
    if (spec === undefined) {
      return fail("crossing-not-granted", `${crossing.channel}:${crossing.direction}@${String(crossing.address)}`);
    }
    const key = canonicalSpec(spec);
    increments.set(key, (increments.get(key) ?? 0) + 1);
  }

  for (const [key, increment] of increments) {
    const current = state.value.counts.get(key) ?? 0;
    if (current > Number.MAX_SAFE_INTEGER - increment) {
      return fail("crossing-count-overflow", key);
    }
  }
  for (const [key, increment] of increments) {
    state.value.counts.set(key, (state.value.counts.get(key) ?? 0) + increment);
  }
  return { ok: true, value: snapshot(state.value) };
}

export function meterCrossingRange(
  grant: ChannelGrant,
  crossing: Omit<ChannelSpec, "startAddress" | "endAddress"> & {
    readonly startAddress: number;
    readonly endAddress: number;
  },
): ChannelGrantResult<ChannelMeterSnapshot> {
  const state = stateFor(grant);
  if (!state.ok) return state;
  if (
    !Number.isInteger(crossing.startAddress) ||
    !Number.isInteger(crossing.endAddress) ||
    crossing.startAddress > crossing.endAddress
  ) {
    return fail(
      "crossing-not-granted",
      `${crossing.channel}:${crossing.direction}@${String(crossing.startAddress)}-${String(crossing.endAddress)}`,
    );
  }

  const increments = new Map<string, number>();
  let address = crossing.startAddress;
  while (address <= crossing.endAddress) {
    const spec = specFor(state.value, crossing.channel, crossing.direction, address);
    if (spec === undefined) {
      return fail("crossing-not-granted", `${crossing.channel}:${crossing.direction}@${String(address)}`);
    }
    const coveredEnd = Math.min(spec.endAddress, crossing.endAddress);
    const key = canonicalSpec(spec);
    increments.set(key, (increments.get(key) ?? 0) + coveredEnd - address + 1);
    address = coveredEnd + 1;
  }

  for (const [key, increment] of increments) {
    const current = state.value.counts.get(key) ?? 0;
    if (current > Number.MAX_SAFE_INTEGER - increment) {
      return fail("crossing-count-overflow", key);
    }
  }
  for (const [key, increment] of increments) {
    state.value.counts.set(key, (state.value.counts.get(key) ?? 0) + increment);
  }
  return { ok: true, value: snapshot(state.value) };
}

File: src/Core/ChannelGrant.fs
namespace Zeta.Core

open System
open System.Globalization

/// Direction in which information crosses a TAS channel boundary.
[<RequireQualifiedAccess>]
type ChannelDirection =
    | Read
    | Write

/// One declared apparatus channel and the inclusive address range it admits.
[<Struct>]
type ChannelSpec =
    { Channel: string
      Direction: ChannelDirection
      StartAddress: int
      EndAddress: int }

/// Typed refusals from channel declaration, grant issuance, and metering.
type ChannelGrantFeedback =
    | EmptyChannelSet
    | InvalidChannelName of channel: string
    | InvalidChannelRange of channel: string * startAddress: int * endAddress: int
    | OverlappingChannelRanges of channel: string * direction: ChannelDirection
    | InvalidExperimenterId of issuedBy: string
    | InvalidDerivedChannelLabel of detail: string
    | RunKeyChannelMismatch of expected: string * actual: string
    | CrossingNotGranted of channel: string * direction: ChannelDirection * address: int
    | CrossingCountOverflow of channel: string * direction: ChannelDirection

[<Struct>]
type ExperimenterId = private ExperimenterId of string

[<RequireQualifiedAccess>]
module ExperimenterId =

    let tryCreate (issuedBy: string) : Result<ExperimenterId, ChannelGrantFeedback> =
        let valid =
            not (String.IsNullOrWhiteSpace issuedBy)
            && issuedBy
               |> Seq.forall (fun c ->
                   (c >= 'a' && c <= 'z')
                   || (c >= 'A' && c <= 'Z')
                   || (c >= '0' && c <= '9')
                   || c = '-'
                   || c = '_'
                   || c = '.')

        if valid then
            Ok(ExperimenterId issuedBy)
        else
            Error(InvalidExperimenterId issuedBy)

    let value (ExperimenterId issuedBy) = issuedBy


/// Validated, canonical channel configuration. Construction rejects ambiguous overlaps.
type ChannelSet = private ChannelSet of ChannelSpec list

[<RequireQualifiedAccess>]
module ChannelSet =

    [<Literal>]
    let private MaxAddress = 0xFFF

    let private directionText direction =
        match direction with
        | ChannelDirection.Read -> "read"
        | ChannelDirection.Write -> "write"

    let private isChannelName (channel: string) =
        not (String.IsNullOrEmpty channel)
        && channel[0] >= 'a'
        && channel[0] <= 'z'
        && channel
           |> Seq.forall (fun c ->
               (c >= 'a' && c <= 'z')
               || (c >= '0' && c <= '9')
               || c = '-')

    let private canonicalSpec (spec: ChannelSpec) =
        String.concat
            ""
            [ spec.Channel
              "-"
              directionText spec.Direction
              "@"
              spec.StartAddress.ToString("x4", CultureInfo.InvariantCulture)
              "-"
              spec.EndAddress.ToString("x4", CultureInfo.InvariantCulture) ]

    let private overlaps (left: ChannelSpec) (right: ChannelSpec) =
        String.Equals(left.Channel, right.Channel, StringComparison.Ordinal)
        && left.Direction = right.Direction
        && left.StartAddress <= right.EndAddress
        && right.StartAddress <= left.EndAddress

    let tryCreate (specs: ChannelSpec list) : Result<ChannelSet, ChannelGrantFeedback> =
        match specs with
        | [] -> Error EmptyChannelSet
        | _ ->
            match specs |> List.tryFind (fun spec -> not (isChannelName spec.Channel)) with
            | Some invalid -> Error(InvalidChannelName invalid.Channel)
            | None ->
                match
                    specs
                    |> List.tryFind (fun spec ->
                        spec.StartAddress < 0
                        || spec.EndAddress > MaxAddress
                        || spec.StartAddress > spec.EndAddress)
                with
                | Some invalid ->
                    Error(InvalidChannelRange(invalid.Channel, invalid.StartAddress, invalid.EndAddress))
                | None ->
                    let indexed = specs |> List.indexed

                    match
                        indexed
                        |> List.tryPick (fun (i, left) ->
                            indexed
                            |> List.tryPick (fun (j, right) ->
                                if i < j && overlaps left right then
                                    Some(left.Channel, left.Direction)
                                else
                                    None))
                    with
                    | Some(channel, direction) -> Error(OverlappingChannelRanges(channel, direction))
                    | None ->
                        specs
                        |> List.sortWith (fun left right ->
                            StringComparer.Ordinal.Compare(canonicalSpec left, canonicalSpec right))
                        |> ChannelSet
                        |> Ok

    let specs (ChannelSet channels) : ChannelSpec list = channels

    let detail (ChannelSet channels) : string =
        channels |> List.map canonicalSpec |> String.concat ","

    /// The channel label that must be carried by the bound run key.
    let runLabel (channels: ChannelSet) : Result<Chip8CrossRunStore.RunChannelLabel, ChannelGrantFeedback> =
        let value = detail channels

        match Chip8CrossRunStore.RunChannelLabel.assisted value with
        | Ok label -> Ok label
        | Error _ -> Error(InvalidDerivedChannelLabel value)


/// Opaque capability issued by the experiment harness and bound to one run key.
[<Sealed>]
type ChannelGrant
    internal
    (
        channels: ChannelSet,
        issuedBy: ExperimenterId,
        runKey: Chip8CrossRunStore.RunKey
    ) =

    member _.Channels = channels
    member _.IssuedBy = issuedBy
    member _.RunKey = runKey


[<RequireQualifiedAccess>]
module internal ChannelGrantHarness =

    /// Issue only when the run key already names this exact apparatus configuration.
    let issue
        (issuedBy: ExperimenterId)
        (runKey: Chip8CrossRunStore.RunKey)
        (channels: ChannelSet)
        : Result<ChannelGrant, ChannelGrantFeedback> =
        match ChannelSet.runLabel channels with
        | Error feedback -> Error feedback
        | Ok expected ->
            let expectedText = Chip8CrossRunStore.RunChannelLabel.value expected
            let actualText = Chip8CrossRunStore.RunChannelLabel.value runKey.ChannelLabel

            if String.Equals(expectedText, actualText, StringComparison.Ordinal) then
                Ok(ChannelGrant(channels, issuedBy, runKey))
            else
                Error(RunKeyChannelMismatch(expectedText, actualText))


type ChannelMeter = private ChannelMeter of Map<ChannelSpec, int64>

type ChannelMeterRow =
    { Channel: string
      Direction: ChannelDirection
      StartAddress: int
      EndAddress: int
      Crossings: int64 }

type ChannelMeterSnapshot =
    { ChannelLabel: string
      IssuedBy: string
      RunKey: string
      Rows: ChannelMeterRow list }

[<RequireQualifiedAccess>]
module ChannelMeter =

    let zero (grant: ChannelGrant) : ChannelMeter =
        grant.Channels
        |> ChannelSet.specs
        |> List.map (fun spec -> spec, 0L)
        |> Map.ofList
        |> ChannelMeter

    let private matchingSpec (grant: ChannelGrant) channel direction address =
        grant.Channels
        |> ChannelSet.specs
        |> List.tryFind (fun spec ->
            String.Equals(spec.Channel, channel, StringComparison.Ordinal)
            && spec.Direction = direction
            && address >= spec.StartAddress
            && address <= spec.EndAddress)

    let cross
        (grant: ChannelGrant)
        (channel: string)
        (direction: ChannelDirection)
        (address: int)
        (ChannelMeter counts)
        : Result<ChannelMeter, ChannelGrantFeedback> =
        match matchingSpec grant channel direction address with
        | None -> Error(CrossingNotGranted(channel, direction, address))
        | Some spec ->
            let current = Map.find spec counts

            if current = Int64.MaxValue then
                Error(CrossingCountOverflow(channel, direction))
            else
                counts |> Map.add spec (current + 1L) |> ChannelMeter |> Ok

    let crossRange
        (grant: ChannelGrant)
        (channel: string)
        (direction: ChannelDirection)
        (startAddress: int)
        (endAddress: int)
        (ChannelMeter initialCounts)
        : Result<ChannelMeter, ChannelGrantFeedback> =
        if startAddress > endAddress then
            Error(CrossingNotGranted(channel, direction, startAddress))
        else
            let rec loop address counts =
                if address > endAddress then
                    Ok(ChannelMeter counts)
                else
                    match matchingSpec grant channel direction address with
                    | None -> Error(CrossingNotGranted(channel, direction, address))
                    | Some spec ->
                        let coveredEnd = min spec.EndAddress endAddress
                        let increment = int64 (coveredEnd - address + 1)
                        let current = Map.find spec counts

                        if current > Int64.MaxValue - increment then
                            Error(CrossingCountOverflow(channel, direction))
                        else
                            loop (coveredEnd + 1) (Map.add spec (current + increment) counts)

            loop startAddress initialCounts

    let snapshot (grant: ChannelGrant) (ChannelMeter counts) : ChannelMeterSnapshot =
        let rows =
            grant.Channels
            |> ChannelSet.specs
            |> List.map (fun spec ->
                { Channel = spec.Channel
                  Direction = spec.Direction
                  StartAddress = spec.StartAddress
                  EndAddress = spec.EndAddress
                  Crossings = Map.find spec counts })

        { ChannelLabel =
            grant.RunKey.ChannelLabel
            |> Chip8CrossRunStore.RunChannelLabel.value
          IssuedBy = ExperimenterId.value grant.IssuedBy
          RunKey = Chip8CrossRunStore.keyText grant.RunKey
          Rows = rows }
